src/assert_matches.rs

```diff
@@ -41,23 +41,43 @@
 /// [`debug_assert_matches!`]: ./macro.debug_assert_matches.html
 #[macro_export]
 macro_rules! assert_matches {
-    ($expression:expr, $( $pattern:pat )|+ $( if $guard: expr )?) => {
+    ($expression:expr, $( $pattern:pat )|+) => {
         match $expression {
-            $( $pattern )|+ $( if $guard )? => {},
+            $( $pattern )|+ => {},
             other => {
                 panic!(r#"assertion failed, expression does not match any of the given variants.
     expression: {:?}
-    variants: {}"#, other, stringify!($($pattern) |+ $(if $guard)?));
+    variants: {}"#, other, stringify!($($pattern) |+));
             }
         }
     };
-    ($expression:expr, $( $pattern:pat )|+ $( if $guard: expr )?, $($arg:tt)+) => {
+    ($expression:expr, $( $pattern:pat )|+ if $guard: expr) => {
```

Comment:
The Kleene operator `?` means at most one repetition ([source](https://doc.rust-lang.org/edition-guide/rust-2018/macros/at-most-once.html)). So it basically can match two cases: either the `if $guard` being present or it being absent. Basically, it's equivalent to writing out two matchers, one to match each case.

A simpler example (from the above source) makes it pretty clear that

```rust
macro_rules! foo {
    ($a:ident, $b:expr) => {
        println!("{}", $a);
        println!("{}", $b);
    };
    ($a:ident) => {
        println!("{}", $a);
    }
}
```

is equivalent to

```rust
macro_rules! foo {
    ($a:ident $(, $b:expr)?) => {
        println!("{}", $a);

        $(
            println!("{}", $b);
         )?
    }
}
```

because the `?` matches the optional presence of `b`. Essentially, what I'm trying to say is that for every Kleene operator used, you could rewrite the same macro with the Kleene operator gone, but would need twice as many matchers.

So in the `assert_matches!` case, the number of matchers doubles. The original matchers were

```rust
($expression:expr, $( $pattern:pat )|+ $( if $guard: expr )?) => ..
($expression:expr, $( $pattern:pat )|+ $( if $guard: expr )?, $($arg:tt)+) => ..
```

and in the equivalent Kleene-less case we end up with twice as many matchers:

```rust
($expression:expr, $( $pattern:pat )|+) => ..
($expression:expr, $( $pattern:pat )|+ if $guard: expr) => ..
($expression:expr, $( $pattern:pat )|+, $($arg:tt)+) => ..
($expression:expr, $( $pattern:pat )|+ if $guard: expr, $($arg:tt)+) => ..
```

Those two code blocks cover all of the same cases :)

So, tl;dr: yes, I believe we can be sure that all cases are fully covered.

Reply:
Alright, thanks for this huge explanation 👍🏼